From f73f6f0ef6d78548ea4ceba3aba8ac2182b91c26 Mon Sep 17 00:00:00 2001
From: Ivan Mironov <mironov.ivan@gmail.com>
Date: Wed, 8 Apr 2026 03:36:41 +0200
Subject: [PATCH 2/2] Fix Zed on Raspberry Pi 5

---
 Cargo.toml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Cargo.toml b/Cargo.toml
index 17efe2180096..ef8c30f29cda 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -782,7 +782,7 @@ which = "6.0.0"
 wasm-bindgen = "0.2.113"
 web-time = "1.1.0"
 webrtc-sys = "0.3.23"
-wgpu = { git = "https://github.com/zed-industries/wgpu.git", branch = "v29" }
+wgpu = { git = "https://github.com/im-0/wgpu.git", branch = "zed-v29-rpi5" }
 windows-core = "0.61"
 yawc = "0.2.5"
 zeroize = "1.8"
-- 
2.53.0

